### PR TITLE
Add Go modules support

### LINKED ISSUE #4234: Migrate from dep to Go Modules
Hi,

I've been working on migrating minikube's current dependency management [from Go dep to Go modules][1], mainly because I like to build projects with `git clone`/`make` and forgetting about `GOPATH`. It ended up simplifying the build infrastructure a bit, so I thought about upstreaming the changes.

The only reference to Go modules I found is [this comment from last September][2] on issue #3037, related to the last `godep` to `dep` migration.

Hence the question: are you interested in this change? If so, I'll prepare a proper pull request (e.g. updating the related documentation).

Regards,
Tiago.

[1]: https://github.com/myhro/minikube/commits/go-mod
[2]: https://github.com/kubernetes/minikube/issues/3037#issuecomment-420875747

### PR DESCRIPTION
**Pull request by myhro**
Moving away from Go dep to Go modules allowed the build infrastructure to be cleaned up a bit. The related docs were also updated.

The whole diff is somewhat long after the `vendor/` directory removal, so I guess reviewing it per commit can make the process a bit easier.

This closes #4234.

### CODE REVIEW

**Comment by afbjorklund** [comment]
"Somewhat" (2,178 files, +323 −929,569 lines)

**Comment by myhro** [comment]
I signed it

**Comment by myhro** [comment]
I signed it

**Comment by tstromberg** [comment]
@minikube-bot OK to test

**Comment by tstromberg** [comment]
This change looks beautiful. Once the integration tests are run, I'll merge it.

**Comment by myhro** [comment]
Thank you for the review. Jenkins build is broken, I'll investigate how to fix it.

**Comment by tstromberg** [comment]
@myhro - We do something funny wand build the kvm2 driver from an old Docker image so that it's built against an older release of Debian for libvirt compatibility. I notice the call currently says:

`which go-bindata || GO111MODULE=off GOBIN=/var/lib/jenkins/go/bin go get github.com/jteeuwen/....`

Let me or @sharifelgamal know if you need any help in getting with getting this sorted out.

**Comment by myhro** [comment]
@tstromberg The `kvm2` driver build system I found quite clever and was able to fix it by adding `git` to its Docker image. `jteeuwen/go-bindata` is a tool I would like to replace, as it is not maintained anymore, but right now I don't have an alternative for it.

Debugging the last failure, I noticed that it was trying to build the `hyperkit` driver under the `karalabe/xgo-1.10.x` image which, as the name suggests, is based on Go `1.10`, a version that doesn't support modules. I've changed it to `karalabe/xgo-1.12.x` and was able to build the driver with:

    MINIKUBE_BUILD_IN_DOCKER=y make out/docker-machine-driver-hyperkit

Now I see that it still failed on Jenkins, so I'll keep digging.

**Comment by myhro** [comment]
@tstromberg Sorry, I only saw that after sending the last comment:

```
Status: Downloaded newer image for karalabe/xgo-1.12.x:latest
docker: Error response from daemon: devmapper: Thin Pool has 148835 free data
blocks which is less than minimum required 163840 free data blocks. Create more
free space in thin pool or use dm.min_free_space option to change behavior.
See 'docker run --help'.
Makefile:282: recipe for target 'out/docker-machine-driver-hyperkit' failed
make: *** [out/docker-machine-driver-hyperkit] Error 125
```

Did the Jenkins builder run out of disk space?

P.s.: I can join Slack if you prefer to avoid the comment ping-pong.

**Comment by tstromberg** [comment]
I found https://stackoverflow.com/questions/47457012/how-to-increase-the-space-data-space-total-used-by-docker-that-value-is-less - which mentioned running "docker system prune". Going to do so on our Jenkins master, which apparently has loads of very old images.

**Comment by tstromberg** [comment]
@minikube-bot OK to test

**Comment by tstromberg** [comment]
It looks like perhaps a merge is in order, since jibber jabber is a recent addition? Jenkins got farther this time: passing the hyperkit build, but failing kvm2:

```
15:16:39 pkg/minikube/constants/constants.go:27:2: cannot find package "github.com/blang/semver" in any of:
15:16:39 	/usr/local/go/src/github.com/blang/semver (from $GOROOT)
15:16:39 	/var/lib/jenkins/go/src/github.com/blang/semver (from $GOPATH)
15:16:39 pkg/minikube/console/console.go:27:2: cannot find package "github.com/cloudfoundry-attic/jibber_jabber" in any of:
...
15:16:39 	/usr/local/go/src/k8s.io/kubernetes/cmd/kubeadm/app/constants (from $GOROOT)
15:16:39 	/var/lib/jenkins/go/src/k8s.io/kubernetes/cmd/kubeadm/app/constants (from $GOPATH)
15:16:39 Makefile:345: recipe for target 'out/docker-machine-driver-kvm2' failed
15:16:39 make: *** [out/docker-machine-driver-kvm2] Error 1
15:16:39 make: *** Waiting for unfinished jobs....
````

**Comment by tstromberg** [comment]
Here's the PR that added jibber jabber 7 days ago: https://github.com/kubernetes/minikube/pull/4199

**Comment by myhro** [comment]
Thanks for fixing the disk space issue. The `out/docker-machine-driver-kvm2` task is failing because Go modules aren't available. As it doesn't runs under Docker, different than the `hyperkit` one, and it's a straight call to `go build`, I suspect that the Jenkins builder has Go `< 1.11` currently installed. I can't confirm that suspicion, as the Go version information isn't available in the build logs.

I'm seeing two possible solutions for this problem:

- We can consider that every driver has to be built under Docker, changing `out/docker-machine-driver-kvm2` to behave like `out/docker-machine-driver-hyperkit`.
- Confirm the Go version hypothesis and, if it's true, update the Jenkins builder to a newer Go version which supports modules.

Which one you prefer?

**Comment by tstromberg** [comment]
go v1.12 is a requirement, as per https://github.com/kubernetes/minikube/blob/master/docs/contributors/build_guide.md

I did notice that the Jenkins master was stuck on go v1.12rc1, so I've upgraded the master to v1.12.5. I'm having similar issues building locally, and I believe the difference between our environments is that Jenkins and myself are building from ~/go/src, which disables go modules by default.

I believe that this PR will need to do is set `env GO111MODULE=on` in the Makefile to guarantee 
 build consistency across environments. I would rather not Docker become a build dependency unless strictly 
 necessary.

File: .travis.yml [hunk 1]
```diff
@@ -3,8 +3,9 @@ os: linux
 sudo: required
 
 go:
-  - 1.x
-go_import_path: k8s.io/minikube
+  - 1.12.x
```
Comment (tstromberg) [inline .travis.yml]: If possible, leave this at 1.x so that we will continue to auto-update when v1.13 comes along.

**Comment by tstromberg** [comment]
Thank you for taking this beast of a PR on, by the way. I just learned today that a tool we would like to use for localization (golang.org/x/text/cmd/gotext) is incompatible with vendor directories, so this PR is perfectly timed. You are a hero =)

**Comment by tstromberg** [comment]
@minikube-bot OK to test

File: .travis.yml (hunk 1 above)
Comment (myhro) [reply .travis.yml]: Done.

**Comment by myhro** [comment]
Thank you for clarifying the version and upgrading the server. I agree that it makes way more sense to use `GO111MODULE=on` and not worry about the import path anymore. I still removed it from the `.travis.yml` file just to be clear that it's not needed.

> I just learned today that a tool we would like to use for localization (golang.org/x/text/cmd/gotext) is incompatible with vendor directories, so this PR is perfectly timed.

I didn't knew that there were projects incompatible with vendor directories. Thanks for the tip. I'm glad that we are taking care of that detail. :-)

I'm a bit confused, as there are no timestamps in the Jenkins logs, but looks like the build worked and the integration tests are running now.

**Comment by tstromberg** [comment]
Looks like the build works great with the most recent changes.

I'll wait for the rest of the integration tests to pass, and then merge it in. Thank you again!

**Comment by tstromberg** [comment]
Huzzah! 🎉 Thank you!

**Comment by myhro** [comment]
That's great. Thank you very much for your support, Thomas. It would be way harder without it.

Hope we can work together again on other features/PRs. :-)

**Pull request merged**